### PR TITLE
[dtensor] add CommDebugMode for debugging

### PR DESCRIPTION
**Pull request by wanchaol**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #114404
* __->__ #113592

This PR adds a CommDebugMode debugging tool to record the number of
distributed collectives, utilizing TorchDispatchMode, the idea borrows
from the FlopCounterMode and we can expand this later to make it more
feature complete like the FlopCounterMode

This is useful for debugging with DTensor and testing, in general this
fits for any complex distributed algorithms where it's non-trival to
understand the algorithm, we can use this tool to understand what
happened under the hood., we can later cover c10d collectives directly

Not sure if it would be a good general distributed debug tool yet,
so adding to the dtensor package first

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @fduwjj @wz337 @tianyu-l @wconstab @yf225 @kiukchung @d4l3k @lucasllc

### CODE REVIEW

**Comment by XilunWu** [comment]
btw, do we have a debug tool on Sharding Propagation which I think can simplify the op strategy development? If not, do we have a plan to add it?

**Comment by wanchaol** [comment]
> btw, do we have a debug tool on Sharding Propagation which I think can simplify the op strategy development? If not, do we have a plan to add it?

that's a different issue and I think that need to be solved via better logging.

File: torch/distributed/_tensor/debug/comm_mode.py [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+from torch.utils._python_dispatch import TorchDispatchMode
+
+
+funcol = torch.ops.c10d_functional
+
+
+class CommDebugMode(TorchDispatchMode):
+    """
+    ``CommDebugMode`` is a context manager that counts the number of
+    communications within its context. It does this using a ``TorchDispatchMode``.
```
Comment (wconstab) [inline torch/distributed/_tensor/debug/comm_mode.py:14]: My primary concern on this PR is summed up by "this docstring is misleading, but I'd rather have the Mode behave like the docstring rather than fix the docstring."  --  it only logs the 3 types of registered functional collectives, not all communications ops.  I can see users being confused why their CommDebugMode didn't capture some collectives when they run it with FSDP for example.

Any thoughts about expanding this to log c10d ops too?  would not be as easy..

Alternatively, make this tool more explicitly focused on DTensor comms and then promise it'll capture all DTensor comms?

File: torch/testing/_internal/distributed/_tensor/common_dtensor.py [hunk 2]
```diff
@@ -81,35 +76,6 @@ def skip_unless_torch_gpu(method: T) -> T:
     return cast(T, skip_if_lt_x_gpu(NUM_DEVICES)(method))
 
 
-@dataclass
-class RedistributeProfile:
-    num_calls: int
-
-
-@contextmanager
-def redistribute_profiler() -> Generator[RedistributeProfile, None, None]:
```
Comment (wconstab) [inline torch/testing/_internal/distributed/_tensor/common_dtensor.py:90]: looks like this is an orthogonal refactor, could stamp it if you split PR

File: torch/distributed/_tensor/debug/comm_mode.py (hunk 1 above)
Comment (wconstab) [reply torch/distributed/_tensor/debug/comm_mode.py:14]: i guess a possible proposal is something like this

- build a func in c++ that marks 'collective called' and hook it up to all the c10d collectives
- bind the 'collective called' to python to also instrument funcol

And have 2 UIs for getting at the data (separately)
- add a 'TORCH_LOGS' mode that just log-prints every collective call, but with no user code edit
- build a 'counter obj' in c++ and bind it out to python, so you can turn on a contextmgr and have it record collectives only during the active range, then progrmatically query info about the collectives (possibly including c++ and py stacktraces of where they got called)

Comment (wanchaol) [reply torch/distributed/_tensor/debug/comm_mode.py:14]: Yeah I was thinking about covering the c10d ops later so that FSDP could also works under this, and also a bulk of missing features currently i.e. identifying module boundaries and print out module hierarchy, etc. let me try out c10d ops to see if is easy to add, if not, I can adjust the docstr to mention that it's currently only supports functional collectives or DTensor comms, where I can add all functional collectives to it. Then in follow up PRs figuring out how the c10d ops should work

Comment (wanchaol) [reply torch/distributed/_tensor/debug/comm_mode.py]: updated the comment to reflect that it only supports functional collectives atm, added a remaining functional collectives.

File: torch/testing/_internal/distributed/_tensor/common_dtensor.py (hunk 2 above)
Comment (wanchaol) [reply torch/testing/_internal/distributed/_tensor/common_dtensor.py]: split it into a PR above :)

**Review by wconstab — APPROVED** [review]
